ENH: add colormap iterating to plot_images

jat255 opened the pull request:
This is something that I've wanted on my own plots for a while, so rather than just coding it up as a one-off, I figured it would be useful to add back to HyperSpy. It meets my requirements, but I'd be happy to get some suggestions of how it could work better.

### Description of the change
Often when plotting images, it can be desired to plot images with different colormaps for each signal (such as when correlating a set of spectra plotted with `plot_spectra`). This PR does that with the following changes:
- Primary change is that the `cmap` option of `plot_images` can now be a list, much like the `color` option of `plot_spectra`
- To go along with this, the `centre_colormap` option has also been changed to be iterable
- An additional option of `mpl_colors` was added to `cmap`, which will loop through the default `matplotlib` colors (useful when comparing line intensities or PCA components/loadings)
- Additionally added a helper utility function called `make_cmap`, which generates a `matplotlib` colormap from a list of colors (this function is used when using the `mpl_colors` option described above

### Progress of the PR
- [x] Change implemented 
- [x] update docstring 
- [x] add tests - **There are no plotting tests currently for `plot_images`, but we should add some**
- [x] update user guide (if appropriate)
- [x] ready for review

### Script I was using locally for testing
```python
import matplotlib as mpl
mpl.use('Qt4Agg')

import hyperspy.api as hs
import matplotlib.pyplot as plt
from hyperspy.drawing.utils import make_cmap
import scipy

plt.ioff()

image = hs.signals.Signal2D([scipy.misc.ascent()]*12)
image2 = hs.signals.Signal2D([scipy.misc.ascent()]*12)
angles = hs.signals.BaseSignal(range(0,120,10))
angles2 = hs.signals.BaseSignal(range(120,0,-10))
image.map(scipy.ndimage.rotate, angle=angles.T, 
	      show_progressbar=False, reshape=False)
image2.map(scipy.ndimage.rotate, angle=angles2.T, 
	       show_progressbar=False, reshape=False)

hs.plot.plot_images(image, per_row=4, 
					suptitle='MPL Colors',
                    cmap='mpl_colors', axes_decor='off',
                    colorbar='multi', tight_layout=True)

hs.plot.plot_images(image, per_row=4, 
					suptitle='Default',
                    cmap=None, axes_decor='off',
                    colorbar='multi', tight_layout=True)

hs.plot.plot_images(image, per_row=4,
					suptitle='One string', 
                    cmap='viridis', axes_decor='off',
                    colorbar='multi', tight_layout=True)

hs.plot.plot_images(image, per_row=4,
					suptitle='Short list', 
                    cmap=['viridis', 'PuBu_r', 'coolwarm',
                    	  'inferno', 'nipy_spectral'], 
                    axes_decor='off',
                    colorbar='multi', tight_layout=True)

hs.plot.plot_images([image, image2], per_row=6,
					suptitle='Two signals', 
                    cmap='mpl_colors', 
                    axes_decor='off',
                    colorbar='multi', tight_layout=True)

hs.plot.plot_images(image, per_row=4,
					suptitle='make_cmap example', 
                    cmap=make_cmap(['#ffbb3e', (0.5,0,0)], 'Custom_cmap'), 
                    axes_decor='off',
                    colorbar='multi', tight_layout=True)

plt.show()
```

jat255 committed:
add make_cmap and enable lists of cmaps in plot_images

jat255 commented:
How it looks when using the script above:

![1-mplcolors](https://user-images.githubusercontent.com/1278301/34631206-569e2e96-f23d-11e7-8dbc-745aad865ccd.png)
![2-default](https://user-images.githubusercontent.com/1278301/34631209-596357e6-f23d-11e7-9dd9-80a23a54c66e.png)
![3-onestring](https://user-images.githubusercontent.com/1278301/34631211-5b0e20da-f23d-11e7-84a5-c3fd47d46068.png)
![4-shortlist](https://user-images.githubusercontent.com/1278301/34631216-5cc36b92-f23d-11e7-88d0-e07d86a7a9dd.png)
![5-twosignals](https://user-images.githubusercontent.com/1278301/34631221-5e5bdcc8-f23d-11e7-8a04-57fde2452c32.png)
![6-makecmap](https://user-images.githubusercontent.com/1278301/34631226-5fead5c6-f23d-11e7-8809-485d75a60f95.png)

jat255 committed:
fix bug in make_cmap bit processing

jat255 committed:
Update docstring for cmap options

jat255 committed:
Add plot_images tests for cmap options

ericpre commented:
For info:

- there are plot testing for `plot_images` in [test_plot_signal2d.py](https://github.com/hyperspy/hyperspy/blob/RELEASE_next_minor/hyperspy/tests/drawing/test_plot_signal2d.py#L118)
- the plot testing in CI are failing because at some point the freetype version changed.

jat255 committed:
Update user guide to include description of new cmap functionality in plot_images

jat255 committed:
Update docstring for plot_images

jat255 committed:
Fix bug with auto-centreing of colormap because of change to itertools.cycle

jat255 commented:
> there are plot testing for plot_images in test_plot_signal2d.py

@ericpre Thanks, I moved my tests into that file

This should be ready for review. All the tests are passing on my machine.

francisco-dlp commented:
This is awesome!

@ericpre, have you reviewed this or would you like me to do it?

ericpre commented:
I can review it a bit later, in case you want to focus on other PRs waiting
for review! ;)

ericpre commented on hyperspy/drawing/utils.py:
```diff
@@ -420,8 +420,16 @@ def plot_images(images,
+            cause the cmap to loop through the default ``matplotlib``
+            colors (to match with the default output of the
+            :py:func:`~.drawing.utils.plot_spectra` method.
+            Note: if using more than one colormap, using the ``'single'``
```
I think that we should not allow to use different colormaps and a single colorbar together. This will quite confusing and I can't think of any situation where it would be useful.

ericpre commented on hyperspy/drawing/utils.py:
```diff
@@ -580,6 +577,46 @@ def plot_images(images,
+        cmap = ['mpl{}'.format(i) for i in
+                    range(len(mpl.rcParams['axes.prop_cycle']))]
+    # cmap is list, tuple, or something else iterable (but not string):
+    elif hasattr(cmap, '__iter__') and not isinstance(cmap, str):
```
You could pass many thing in there and get an error later in the code. I would rather consider all the expected input and finished in the `else` by raising a ValueError when the parameter is not correct to provide a clear error message.

ericpre commented on doc/user_guide/visualisation.rst:
```diff
@@ -548,8 +548,92 @@ which is used to call subplots_adjust method of matplotlib
+
+.. NOTE::
+
+    If a list or other iterable type is given for the ``cmap`` argument
```
See comments below.

ericpre reviewed (changes requested):
Looks good, just a few comments.
I suspect you also need to merge with upstream to update the other baselines images since we have to change to freetype 2.8...

jat255 committed:
Disallow using single colorbar when using multiple colormaps; update docstrings to reflect this; raise warning when using a bogus colormap value.

jat255 committed:
Update user guide

jat255 committed:
Move plot_images tests into plot_signal2d; update and add plot test images

jat255 commented:
Ok, sorry for the delay, but this has been updated as requested. I rebased onto RNMinor, updated to freetype 2.8 locally, and regenerated the associated plots, so I think this should be ready to go.

jat255 commented:
Actually, hold off on this. It looks like I might have made a license violation by using the ``make_cmap`` code I found online. It's licensed under a [CC BY-NC-ND 3.0](https://creativecommons.org/licenses/by-nc-nd/3.0/deed.en_US) which is not open-source, and does not allow derivative code, so I don't think it would be compatible with HyperSpy. I've reached out to the author of that code for an exemption, and I'll update here when he gets back to me.

jat255 commented:
The license thing is no longer an issue. The author got back to me, and he has [published the code](https://github.com/CSlocumWX/custom_colormap/blob/master/custom_colormaps.py) under a BSD-style license, so we should not have a problem using it.

jat255 committed:
Update license information for borrowed code

jat255 committed:
Update plot_images to raise ValueError for incompatible arguments; update tests to match

jat255 commented:
Alright, should be good to go and ready for review once more. Per @francisco-dlp's suggestion, I updated the code to raise an error when incompatible arguments are given, and the tests have been updated to check for this.

jat255 commented:
@ericpre any chance you could look this over if you have some time?

ericpre commented on hyperspy/drawing/utils.py:
```diff
@@ -580,6 +582,52 @@ def plot_images(images,
+            cmap = [c.name for c in cmap]  # convert colormap to string
+        except AttributeError:
+            cmap = [c for c in cmap]   # c should be string if not colormap
+        colorbar = __check_single_colorbar(colorbar)
```
same as above.

ericpre commented on hyperspy/drawing/utils.py:
```diff
@@ -580,6 +582,52 @@ def plot_images(images,
+                      name='mpl{}'.format(n_color))
+        cmap = ['mpl{}'.format(i) for i in
+                range(len(mpl.rcParams['axes.prop_cycle']))]
+        colorbar = __check_single_colorbar(colorbar)
```
`__check_single_colorbar(colorbar)` return `None` in any case and it then removes the colorbar from the plot, by setting `colorbar=None`!!! I would just do the check and not the assign the result of the function to `colorbar`.

ericpre reviewed (commented):
This looks good, just one small bug to fix!

ericpre committed:
Fix bug with missing colorbar and update baseline images.

ericpre commented:
I have fixed the issue with colorbar and updated the baseline images accordingly.

jat255 commented:
@ericpre thanks for fixing that bug. I was out of town on travel for a few days and so couldn't get to it. Glad to see this merged in!
